ocaml/xapi/xapi_pool_update.ml

```diff
@@ -532,21 +532,50 @@ let resync_host ~__context ~host =
     |> List.iter (fun self -> destroy ~__context ~self)
   end
 
-let path_from_uri uri =
+let path_and_host_from_uri uri =
   (* remove any dodgy use of "." or ".." NB we don't prevent the use of symlinks *)
-  String.sub_to_end uri (String.length Constants.get_pool_update_download_uri)
-  |> Filename.concat !Xapi_globs.host_update_dir
-  |> Uri.pct_decode
-  |> Stdext.Unixext.resolve_dot_and_dotdot
+  let host_and_path = String.sub_to_end uri (String.length Constants.get_pool_update_download_uri) in
+  match String.split ~limit:2 '/' host_and_path with
+  | [host; untrusted_path] ->
+    let resolved_path = untrusted_path
+      |> Filename.concat !Xapi_globs.host_update_dir
+      |> Uri.pct_decode
+      |> Stdext.Unixext.resolve_dot_and_dotdot
+    in
+    (host,resolved_path)
+  | _ -> failwith "Expecting both host and path in the uri"
+
+let proxy_request req s host_uuid =
+  Server_helpers.exec_with_new_task "pool_update_proxy_request" (fun __context ->
+    let host =
+      try Some (Db.Host.get_by_uuid ~__context ~uuid:host_uuid) with _ -> None
+    in
+    match host with
+    |Some host ->
+      let ip = Db.Host.get_address ~__context ~self:host in
+      let transport = Xmlrpc_client.(SSL(SSL.make (), ip, 443)) in
+      Xmlrpc_client.with_transport transport (fun fd ->
+        Unixext.really_write_string fd (Http.Request.to_wire_string req);
+        Unixext.proxy s fd)
+    |None ->
+      debug "Caught exception while get Host by uuid %s" host_uuid;
+      Http_svr.response_badrequest ~req s;
+    req.Request.close <- true
```

Comment:
In spite of the indentation, this line is part of the `| None ->` case (only). I think that rather than having it here, we should put this line near the top of `pool_update_download_handler`, because we always want to handle just one request, not matter what the response will be, so we never want to keep the connection alive after the response.

Even before this patch, we don't actually close the connection on the path that calls `Http_svr.response_forbidden`, which looks wrong to me. Moving the line to the top would fix that as well.

Searching for `close <- true` reveals that many similar handlers do it this way.

Reply:
For both 'forbidden' and 'badrequest' we instruct the client to close the connection, which should cause us to close too (see https://github.com/xapi-project/xen-api-libs-transitional/blob/master/http-svr/http.ml#L30-L33 and https://github.com/xapi-project/xen-api-libs-transitional/blob/master/http-svr/http.ml#L57-L60). However, it's good form for us to close proactively so we should set `req.Request.close` to true in all cases, as @robhoes suggests.

Reply:
Note that the we may end up handling multiple requests on the one connection - since after the initial headers we proxy all the traffic, any subsequent requests will also be proxied, until either the client or remote server closes the connection.